fortuna/hallucination/base.py

```diff
@@ -0,0 +1,335 @@
+from typing import (
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+import numpy as np
+from sklearn.manifold import locally_linear_embedding
+from sklearn.mixture import GaussianMixture
+import torch
+from torch import nn
+from tqdm import tqdm
+from transformers import PreTrainedTokenizer
+
+from fortuna.conformal import BinaryClassificationMulticalibrator
+from fortuna.hallucination.grouping.clustering.base import GroupingModel
+from fortuna.hallucination.scoring.inv_perplexity import inv_perplexity
+
+
+class HallucinationMulticalibrator:
+    def __init__(
+        self,
+        generative_model: nn.Module,
+        tokenizer: PreTrainedTokenizer,
+        embedding_reduction_fn: Callable[[np.ndarray], np.ndarray] = None,
+        clustering_models: Optional[List] = None,
+        scoring_fn: Callable[
```

Comment:
embedding_reduction_fn and scoring_fn deal differently with default settings. why?